src/Compilers/CSharp/Portable/Binder/Semantics/BestTypeInferrer.cs

```diff
@@ -258,7 +246,10 @@ private static TypeSymbol Better(TypeSymbol type1, TypeSymbol type2, Conversions
 
                 if (type1.Equals(type2, TypeCompareKind.IgnoreDynamicAndTupleNames))
                 {
-                    return MethodTypeInferrer.Merge(type1, type2, conversions.CorLibrary);
+                    var type1WithAnnotations = TypeSymbolWithAnnotations.Create(type1);
+                    var type2WithAnnotations = TypeSymbolWithAnnotations.Create(type2);
+                    (bool success, TypeSymbolWithAnnotations merged) = MethodTypeInferrer.TryMerge(type1WithAnnotations, type2WithAnnotations, VarianceKind.Out, conversions);
```

Comment:
>TryMerge [](start = 90, length = 8)

With the merging rules we came up with, it should not be possible for this to fail. However, it should be possible to want to report a diagnostic, which we could either place into useSiteDiagnostics or return a boolean flag for. Either way, this should always compute a non-null result type, and we should assert that.

Reply:
Added an assert to `TypeSymbolWithAnnotations.MergeNullability`.

---
In reply to: [224636425](https://github.com/dotnet/roslyn/pull/29727#discussion_r224636425) [](ancestors = 224636425)